### PR TITLE
Critical Section __m512i

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
Hello, this is about 75% of cpu time in ethash_hash.
The code does not yet function, but I would like to start a discussion.

**Comment by bobsummerwill** [comment]
I'll see if I can loop some appropriate people in, @mancoast.   Thanks!

**Comment by bobsummerwill** [comment]
What is **MIC**, please, @mancoast?
The only online references I can find are in Intel docs talking about "co-processor", but without any more detail than that.

**Comment by bobsummerwill** [comment]
Is this "Xeon Phi Coproc", again?
Same as https://github.com/ethereum/libweb3core/pull/71?

**Comment by ghost** [comment]
Greetings @bobsummerwill,

MIC is an acronym for many integrate core architecture aka Xeon Phi Coproc.

File: libethash/internal.c [hunk 1]
```diff
@@ -208,7 +219,14 @@ static bool ethash_hash(
 		for (unsigned n = 0; n != MIX_NODES; ++n) {
 			node const* dag_node;
 			if (full_nodes) {
+#if defined(__MIC__)
+				// TODO: fix dag_node
+				node tmp_node;
+				tmp_node = full_nodes[MIX_NODES * index + n];
+				dag_node = &tmp_node;
+#else
 				dag_node = &full_nodes[MIX_NODES * index + n];
```
Comment (ghost) [inline libethash/internal.c:222]: @bobsummerwill @chriseth 
Hello,

Any idea where memory allocation for the full_nodes object occurs?
It seems that the memory is not aligned for use with m512i.
The code produces a Segmentation Fault without a tmp_node copy.
A Segmentation Fault usually indicates that the memory is not aligned.

Thanks,
coast

**Comment by ghost** [comment]
These changes offer a %100 speed increase using the k1om 5110P.

**Comment by bobsummerwill** [comment]
@mancoast Please could you pull from develop to address the Ubuntu build break.

@chriseth Thoughts on merging?   Thanks!

**Comment by chriseth** [comment]
I would like to have some more tests before this is merged, so I think we should wait for our automation infrastructure transition.

**Comment by ghost** [comment]
Sorry, I haven't had time to revisit these commits.
Work in progress.

**Pull request closed without merging**